# NovaTel Customer Retention & Revenue Intelligence System

## Data Audit

### Objective

The objective of this notebook is to assess the quality, completeness, consistency, and reliability of the dataset before performing any analysis.

### Key Audit Areas

- Dataset Structure
- Data Types
- Missing Values
- Duplicate Records
- Unique Value Analysis
- Categorical Validation
- Numerical Validation
- Business Rule Validation

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
df = pd.read_csv("../data/raw/telecom_customer_churn.csv")

df.head()

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
0,0002-ORFBO,Female,37,Yes,0,Frazier Park,93225,34.827662,-118.999073,2,9,NaN,Yes,42.39,No,Yes,Cable,16.0,No,Yes,No,Yes,Yes,No,No,Yes,One Year,Yes,Credit Card,65.6,593.30,0.00,0,381.51,974.81,Stayed,NaN,NaN
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,NaN,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
2,0004-TLHLJ,Male,50,No,0,Costa Mesa,92627,33.645672,-117.922613,0,4,Offer E,Yes,33.65,No,Yes,Fiber Optic,30.0,No,No,Yes,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,73.9,280.85,0.00,0,134.60,415.45,Churned,Competitor,Competitor had better devices
3,0011-IGKFF,Male,78,Yes,0,Martinez,94553,38.014457,-122.115432,1,13,Offer D,Yes,27.82,No,Yes,Fiber Optic,4.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.0,1237.85,0.00,0,361.66,1599.51,Churned,Dissatisfaction,Product dissatisfaction
4,0013-EXCHZ,Female,75,Yes,0,Camarillo,93010,34.227846,-119.079903,3,3,NaN,Yes,7.38,No,Yes,Fiber Optic,11.0,No,No,No,Yes,Yes,No,No,Yes,Month-to-Month,Yes,Credit Card,83.9,267.40,0.00,0,22.14,289.54,Churned,Dissatisfaction,Network reliability


## 1. Dataset Structure

### Purpose

Understand the overall structure of the dataset before performing data quality checks.

### Questions

- How many rows exist?
- How many columns exist?
- What columns are available?
- Does the dataset match business expectations?

In [3]:
print("Dataset Shape:", df.shape)

print("\nColumns:\n")

for col in df.columns:
    print(col)

Dataset Shape: (7043, 38)

Columns:

Customer ID
Gender
Age
Married
Number of Dependents
City
Zip Code
Latitude
Longitude
Number of Referrals
Tenure in Months
Offer
Phone Service
Avg Monthly Long Distance Charges
Multiple Lines
Internet Service
Internet Type
Avg Monthly GB Download
Online Security
Online Backup
Device Protection Plan
Premium Tech Support
Streaming TV
Streaming Movies
Streaming Music
Unlimited Data
Contract
Paperless Billing
Payment Method
Monthly Charge
Total Charges
Total Refunds
Total Extra Data Charges
Total Long Distance Charges
Total Revenue
Customer Status
Churn Category
Churn Reason


### Findings

- Dataset contains 7,043 customer records and 38 columns.
- Dataset is customer-granular (one row per customer).
- Dataset includes demographic, geographic, service, contract, revenue, and churn information.
- Dataset appears suitable for churn, retention, loyalty, product, and revenue analysis.

## 2. Data Type Validation

### Purpose

Validate whether each column has the correct data type for analysis.

### Questions

- Which columns are numeric?
- Which columns are categorical?
- Are any columns using incorrect data types?

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   str    
 1   Gender                             7043 non-null   str    
 2   Age                                7043 non-null   int64  
 3   Married                            7043 non-null   str    
 4   Number of Dependents               7043 non-null   int64  
 5   City                               7043 non-null   str    
 6   Zip Code                           7043 non-null   int64  
 7   Latitude                           7043 non-null   float64
 8   Longitude                          7043 non-null   float64
 9   Number of Referrals                7043 non-null   int64  
 10  Tenure in Months                   7043 non-null   int64  
 11  Offer                              3166 non-null   str    
 12  Pho

### Findings

- Data types appear appropriate for analysis.
- No immediate type conversion issues identified.
- Several columns contain null values and require business validation.
- Potential business-driven nulls are present in service and churn-related fields.

## 3. Missing Value Analysis

### Purpose

Identify missing values and determine whether they represent data quality issues or valid business scenarios.

In [5]:
missing = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing %": round((df.isnull().sum() / len(df)) * 100, 2)
})

missing = missing[missing["Missing Count"] > 0]

missing.sort_values(
    by="Missing Count",
    ascending=False
)

,Missing Count,Missing %
Churn Category,5174,73.46
Churn Reason,5174,73.46
Offer,3877,55.05
Online Security,1526,21.67
Online Backup,1526,21.67
Avg Monthly GB Download,1526,21.67
Internet Type,1526,21.67
Streaming Movies,1526,21.67
Streaming TV,1526,21.67
Device Protection Plan,1526,21.67


### Validation of Internet Service Related Nulls

Determine whether internet-related null values occur because customers do not subscribe to internet service.

In [6]:
df[df["Internet Type"].isnull()]["Internet Service"].value_counts(dropna=False)

Internet Service
No    1526
Name: count, dtype: int64

### Validation of Phone Service Related Nulls

Determine whether phone-related null values occur because customers do not subscribe to phone service.

In [7]:
df[df["Multiple Lines"].isnull()]["Phone Service"].value_counts(dropna=False)

Phone Service
No    682
Name: count, dtype: int64

### Validation of Churn Related Nulls

Determine whether churn-related null values occur because customers have not churned.

In [8]:
df[df["Churn Category"].isnull()]["Customer Status"].value_counts(dropna=False)

Customer Status
Stayed    4720
Joined     454
Name: count, dtype: int64

### Missing Value Findings

No evidence of data quality issues has been identified from missing values.

Business-driven null values were found in:

- Internet-related columns for customers without internet service.
- Multiple Lines for customers without phone service.
- Churn Category and Churn Reason for customers who did not churn.

These null values represent valid business scenarios and should not be treated as missing data errors.

## 4. Duplicate Record Validation

### Purpose

Verify whether Customer ID uniquely identifies each customer.

In [9]:
print("Total Rows:", len(df))

print("Unique Customer IDs:", df["Customer ID"].nunique())

Total Rows: 7043
Unique Customer IDs: 7043


### Findings

- Customer ID uniquely identifies each record.
- Number of unique Customer IDs equals the total number of rows.
- No duplicate customer records were identified.
- Duplicate removal is not required.

## 5. Empty Record Validation

### Purpose

Determine whether any rows contain entirely missing information and should be removed.

In [10]:
df.isnull().all(axis=1).sum()

np.int64(0)

### Findings

- No completely empty records were found.
- Dataset contains valid information for all customer records.
- No row removal is required.

## 6. Categorical Data Validation

### Purpose

Validate categorical fields for:

- Invalid values
- Unexpected categories
- Missing categories
- Business consistency

The following analysis begins with the primary target variable: Customer Status.

### Customer Status Validation

Validate the target variable used for churn and retention analysis.

In [11]:
df["Customer Status"].value_counts(dropna=False)

Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64

### Customer Status Findings

- Three valid customer states were identified: Stayed, Churned, and Joined.
- No null values were found.
- No unexpected category values were identified.
- Customer Status will serve as the primary business outcome variable for churn and retention analysis.

### Contract Validation

Validate contract categories and identify available contract types.

In [12]:
df["Contract"].value_counts(dropna=False)

Contract
Month-to-Month    3610
Two Year          1883
One Year          1550
Name: count, dtype: int64

### Contract Findings

- Three valid contract categories were identified:
  - Month-to-Month
  - One Year
  - Two Year
- No null values were found.
- No invalid or unexpected category values were detected.
- The Contract column is suitable for retention and churn analysis.

### Internet Service Validation

Validate internet service categories and identify available customer internet service options.

In [13]:
df["Internet Service"].value_counts(dropna=False)

Internet Service
Yes    5517
No     1526
Name: count, dtype: int64

### Internet Service Findings

- Two valid categories were identified:
  - Yes
  - No
- No null values were found.
- No invalid category values were detected.
- Customers without internet service account for the internet-related null values identified earlier.
- These null values represent valid business scenarios rather than data quality issues.

### Internet Type Validation

Validate available internet technologies offered to customers.

In [14]:
df["Internet Type"].value_counts(dropna=False)

Internet Type
Fiber Optic    3035
DSL            1652
NaN            1526
Cable           830
Name: count, dtype: int64

### Internet Type Findings

- Three internet technologies were identified:
  - Fiber Optic
  - DSL
  - Cable
- Null values correspond to customers without internet service.
- No invalid category values were detected.
- Internet Type is suitable for product and churn analysis.

### Offer Validation

Validate marketing offer categories and identify offer distribution across customers.

In [15]:
df["Offer"].value_counts(dropna=False)

Offer
NaN        3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64

### Offer Findings

- Five offer categories were identified:
  - Offer A
  - Offer B
  - Offer C
  - Offer D
  - Offer E
- No invalid offer categories were detected.
- A large number of records contain null values.
- Additional validation is required to determine whether these null values represent a business scenario or a data quality issue.

### Offer Null Validation

Determine whether customers with null offers represent a valid business scenario.

In [16]:
df[df["Offer"].isnull()]["Customer Status"].value_counts(dropna=False)

Customer Status
Stayed     2547
Churned    1051
Joined      279
Name: count, dtype: int64

### Offer Null Validation Findings

- Null offer values occur across Stayed, Churned, and Joined customers.
- Current evidence does not explain the business meaning of null offer values.
- No data quality issue has been proven.
- The business interpretation of null offer values remains unresolved and should be documented as an assumption if used later in analysis.

### Remaining Categorical Columns Validation

Validate category consistency across remaining business fields.

In [17]:
categorical_cols = [
    "Gender",
    "Married",
    "Phone Service",
    "Paperless Billing",
    "Payment Method"
]

for col in categorical_cols:
    print(f"\n{'='*50}")
    print(col)
    print(df[col].value_counts(dropna=False))


Gender
Gender
Male      3555
Female    3488
Name: count, dtype: int64

Married
Married
No     3641
Yes    3402
Name: count, dtype: int64

Phone Service
Phone Service
Yes    6361
No      682
Name: count, dtype: int64

Paperless Billing
Paperless Billing
Yes    4171
No     2872
Name: count, dtype: int64

Payment Method
Payment Method
Bank Withdrawal    3909
Credit Card        2749
Mailed Check        385
Name: count, dtype: int64


### Customer & Billing Fields Findings

Validated columns:

- Gender
- Married
- Phone Service
- Paperless Billing
- Payment Method

Findings:

- No null values were identified.
- No unexpected category values were detected.
- No spelling inconsistencies were observed.
- All columns are suitable for business analysis.

In [18]:
service_cols = [
    "Online Security",
    "Online Backup",
    "Device Protection Plan",
    "Premium Tech Support",
    "Streaming TV",
    "Streaming Movies",
    "Streaming Music",
    "Unlimited Data"
]

for col in service_cols:
    print(f"\n{'='*50}")
    print(col)
    print(df[col].value_counts(dropna=False))


Online Security
Online Security
No     3498
Yes    2019
NaN    1526
Name: count, dtype: int64

Online Backup
Online Backup
No     3088
Yes    2429
NaN    1526
Name: count, dtype: int64

Device Protection Plan
Device Protection Plan
No     3095
Yes    2422
NaN    1526
Name: count, dtype: int64

Premium Tech Support
Premium Tech Support
No     3473
Yes    2044
NaN    1526
Name: count, dtype: int64

Streaming TV
Streaming TV
No     2810
Yes    2707
NaN    1526
Name: count, dtype: int64

Streaming Movies
Streaming Movies
No     2785
Yes    2732
NaN    1526
Name: count, dtype: int64

Streaming Music
Streaming Music
No     3029
Yes    2488
NaN    1526
Name: count, dtype: int64

Unlimited Data
Unlimited Data
Yes    4745
NaN    1526
No      772
Name: count, dtype: int64


### Internet Service Feature Findings

Validated columns:

- Online Security
- Online Backup
- Device Protection Plan
- Premium Tech Support
- Streaming TV
- Streaming Movies
- Streaming Music
- Unlimited Data

Findings:

- All columns contain expected category values.
- Null values are explained by customers without internet service.
- No invalid categories were detected.
- No data quality issues were identified.

## 7. Numerical Data Validation

### Purpose

Validate numerical fields for:

- Unexpected values
- Negative values
- Outliers requiring investigation
- Business rule consistency

In [19]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns

df[numeric_cols].describe().T

,count,mean,std,min,25%,50%,75%,max
Age,7043.0,46.509726,16.750352,19.000000,32.000000,46.000000,60.000000,80.000000
Number of Dependents,7043.0,0.468692,0.962802,0.000000,0.000000,0.000000,0.000000,9.000000
Zip Code,7043.0,93486.070567,1856.767505,90001.000000,92101.000000,93518.000000,95329.000000,96150.000000
Latitude,7043.0,36.197455,2.468929,32.555828,33.990646,36.205465,38.161321,41.962127
Longitude,7043.0,-119.756684,2.154425,-124.301372,-121.788090,-119.595293,-117.969795,-114.192901
Number of Referrals,7043.0,1.951867,3.001199,0.000000,0.000000,0.000000,3.000000,11.000000
Tenure in Months,7043.0,32.386767,24.542061,1.000000,9.000000,29.000000,55.000000,72.000000
Avg Monthly Long Distance Charges,6361.0,25.420517,14.200374,1.010000,13.050000,25.690000,37.680000,49.990000
Avg Monthly GB Download,5517.0,26.189958,19.586585,2.000000,13.000000,21.000000,30.000000,85.000000
Monthly Charge,7043.0,63.596131,31.204743,-10.000000,30.400000,70.050000,89.750000,118.750000


### Numerical Validation Findings

- Most numerical fields fall within reasonable business ranges.
- No obvious issues were identified in demographic, geographic, tenure, referral, or revenue-related fields.
- A potential anomaly was detected in Monthly Charge where the minimum value is negative.
- Additional validation is required before determining whether this represents a data quality issue or a valid business scenario.

### Monthly Charge Anomaly Investigation

Investigate customers with negative monthly charges.

In [20]:
df[df["Monthly Charge"] < 0]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,NaN,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
32,0052-DCKON,Male,28,Yes,0,Los Angeles,90063,34.044271,-118.185237,9,66,NaN,Yes,43.77,Yes,Yes,Fiber Optic,41.0,Yes,Yes,Yes,Yes,Yes,Yes,Yes,Yes,One Year,No,Bank Withdrawal,-2.0,7942.15,0.00,0,2888.82,10830.97,Stayed,NaN,NaN
170,0274-JKUJR,Male,21,Yes,2,Los Angeles,90040,33.994524,-118.149953,9,7,Offer E,No,NaN,NaN,Yes,Cable,47.0,Yes,Yes,Yes,No,Yes,Yes,Yes,Yes,Month-to-Month,No,Credit Card,-3.0,465.70,0.00,0,0.00,465.70,Stayed,NaN,NaN
232,0361-HJRDX,Female,38,No,0,Sherman Oaks,91403,34.147149,-118.463365,0,68,Offer A,No,NaN,NaN,Yes,DSL,28.0,Yes,Yes,Yes,Yes,Yes,Yes,Yes,No,Two Year,No,Credit Card,-8.0,4539.60,0.00,130,0.00,4669.60,Stayed,NaN,NaN
336,0495-ZBNGW,Male,78,Yes,0,Loma Mar,94021,37.266388,-122.263080,1,5,NaN,Yes,6.66,Yes,Yes,Fiber Optic,14.0,No,No,Yes,No,No,No,No,No,Month-to-Month,Yes,Bank Withdrawal,-1.0,343.95,0.00,20,33.30,397.25,Churned,Competitor,Competitor offered more data
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6894,9793-WECQC,Male,53,No,0,Foresthill,95631,39.031876,-120.811141,0,23,NaN,Yes,48.21,No,Yes,Fiber Optic,23.0,No,Yes,No,No,Yes,Yes,Yes,Yes,Month-to-Month,No,Credit Card,-1.0,2192.90,0.00,0,1108.83,3301.73,Stayed,NaN,NaN
6918,9821-BESNZ,Male,43,No,0,Northridge,91325,34.236683,-118.517588,0,66,Offer A,Yes,33.02,Yes,Yes,Fiber Optic,6.0,No,Yes,No,Yes,Yes,Yes,Yes,Yes,One Year,Yes,Bank Withdrawal,-7.0,6860.60,44.00,0,2179.32,8995.92,Stayed,NaN,NaN
6976,9898-KZQDZ,Female,65,Yes,0,Lee Vining,93541,37.890145,-119.184087,1,40,NaN,Yes,49.50,Yes,Yes,Fiber Optic,20.0,No,No,No,No,Yes,Yes,Yes,Yes,Month-to-Month,Yes,Credit Card,-1.0,3640.45,0.00,0,1980.00,5620.45,Churned,Competitor,Competitor had better devices
7020,9961-JBNMK,Male,79,No,0,Desert Center,92239,33.889605,-115.257009,0,21,NaN,Yes,42.27,No,Yes,Fiber Optic,7.0,No,No,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,-1.0,2030.30,10.20,0,887.67,2907.77,Churned,Dissatisfaction,Product dissatisfaction


### Monthly Charge Investigation Findings

- 120 customer records contain negative Monthly Charge values.
- The negative values appear across multiple customer segments.
- The pattern suggests a systematic business process rather than isolated data entry errors.
- The business meaning of these values cannot be determined from current evidence.
- Negative Monthly Charge values should be retained for further investigation during the Data Cleaning phase.

### Revenue Field Validation

Check revenue-related fields for unexpected negative values.

In [21]:
revenue_cols = [
    "Total Charges",
    "Total Refunds",
    "Total Extra Data Charges",
    "Total Long Distance Charges",
    "Total Revenue"
]

for col in revenue_cols:
    print(col)
    print("Min:", df[col].min())
    print("-" * 30)

Total Charges
Min: 18.8
------------------------------
Total Refunds
Min: 0.0
------------------------------
Total Extra Data Charges
Min: 0
------------------------------
Total Long Distance Charges
Min: 0.0
------------------------------
Total Revenue
Min: 21.36
------------------------------


### Revenue Field Findings

- Revenue-related fields contain valid value ranges.
- No unexpected negative values were detected.
- Monthly Charge remains the only field requiring further investigation.
- No immediate revenue data quality issues were identified.

## Data Dictionary Inspection

### Purpose

Understand the meaning of dataset fields and business definitions provided by the source.

In [23]:
data_dict = pd.read_csv(
    "../data/raw/telecom_data_dictionary.csv",
    encoding="latin1"
)

print("Shape:", data_dict.shape)

data_dict.head()

Shape: (40, 3)


,Table,Field,Description
0,Customer Churn,CustomerID,A unique ID that identifies each customer
1,Customer Churn,Gender,"The customer’s gender: Male, Female"
2,Customer Churn,Age,"The customer’s current age, in years, at the t..."
3,Customer Churn,Married,"Indicates if the customer is married: Yes, No"
4,Customer Churn,Number of Dependents,Indicates the number of dependents that live w...


In [24]:
zipcode_df = pd.read_csv(
    "../data/raw/telecom_zipcode_population.csv"
)

print("Shape:", zipcode_df.shape)

zipcode_df.head()

Shape: (1671, 2)


,Zip Code,Population
0,90001,54492
1,90002,44586
2,90003,58198
3,90004,67852
4,90005,43019


# Audit Summary

## Files Reviewed

### 1. telecom_customer_churn.csv

- 7,043 customer records
- 38 columns
- Customer-level dataset
- Primary dataset used for analysis, SQL, and dashboard development

### 2. telecom_data_dictionary.csv

- 40 records
- 3 columns
- Reference file containing field definitions and business descriptions
- Used for documentation and business understanding

### 3. telecom_zipcode_population.csv

- 1,671 records
- 2 columns
- Lookup table containing population information by zip code
- Expected to support geographic analysis and data modeling

---

## Key Findings

### Dataset Structure

- Customer ID uniquely identifies each customer.
- No duplicate customer records were identified.
- No completely empty records were found.
- Dataset structure aligns with the business problem and project objectives.

### Missing Values

Most missing values were determined to be valid business scenarios rather than data quality issues.

Examples include:

- Internet-related fields are null when Internet Service = No.
- Multiple Lines is null when Phone Service = No.
- Churn Category and Churn Reason are null for customers who did not churn.

### Data Types

- Data types are appropriate for analysis.
- No major data type issues were identified.

### Categorical Fields

- No unexpected category values were detected.
- No spelling inconsistencies were identified.
- All major business categories appear valid and usable for analysis.

### Numerical Fields

- Numerical fields contain reasonable business values.
- One anomaly was identified:
  - Monthly Charge contains 120 negative values.
- These values appear systematic and require further investigation during data cleaning.

### Offer Field

- Offer contains a large number of null values.
- Current evidence does not explain the business meaning of these null values.
- Additional business interpretation may be required during analysis.

---

## Audit Outcome

### Dataset Quality Assessment

- Dataset quality is good.
- Dataset is suitable for business analysis, SQL analysis, KPI development, and dashboard creation.
- No major structural issues were identified.

### Items for Data Cleaning Review

1. Investigate negative Monthly Charge values.
2. Determine business interpretation of null values in Offer.

---

## Conclusion

The NovaTel datasets are suitable for analysis and project development. A light audit was performed to validate structure, data quality, missing values, categorical fields, numerical fields, and supporting files. The datasets are ready for the Data Cleaning phase.